mitodl/mit-open
# fix sorting

### What are the relevant tickets?
closes https://github.com/mitodl/hq/issues/4097
closes https://github.com/mitodl/hq/issues/4121
closes https://github.com/mitodl/hq/issues/4119
closes https://github.com/mitodl/hq/issues/4117

### Description (What does it do?)
This pr 
1) Adds an 'upcoming' sort for api/v1/learning_resources_search and refactors 'upcoming' for api/v1/learning_resources
The upcoming sort needs to search through the runs to find the run with the soonest future date before sorting. It's more efficient to save this value instead of calculating it on the fly. There is a task that will run at midnight that will update any next_start_dates that are in the past

2) Removes the "Extra actions" filters  for api/v1/learning_resources since those duplicate sorts and are confusing

3) Fixes bugs that caused the existing /upcoming filter and sortby=views (but not the filter) to return duplicate results

4) Adds created_on to the data returned by the serializer since it's a possible sort value and it's useful to be able to see if the sort is working. This is temporary https://github.com/mitodl/hq/issues/4127  - we should have a better published date to use for sort in the future

### Screenshots (if appropriate):


### How can this be tested?
- Run  run web ./manage.py migrate learning_resources (if it doesn't run automatically)
- Verify that next_start_date is populated for learning resources that have runs in the future
- Go to http://localhost:8063/admin/learning_resources/learningresource/ and set the next_start_date value for a learning resource to a value in the past
- Run
```
from learning_resources.tasks import *
update_next_start_date()
```
- Verify that the value is updated to either the soonest upcoming run date or null

- Run docker-compose run web ./manage.py recreate_index --all
- Verify that the following api calls work as expected
http://localhost:8063/api/v1/learning_resources/?sortby=-created_on
http://localhost:8063/api/v1/learning_resources/?sortby=created_on
http://localhost:8063/api/v1/learning_resources_search/?sortby=-created_on
http://localhost:8063/api/v1/learning_resources_search/?sortby=created_on
http://localhost:8063/api/v1/learning_resources/?sortby=views
http://localhost:8063/api/v1/learning_resources/?sortby=-views
http://localhost:8063/api/v1/learning_resources_search/?sortby=views
http://localhost:8063/api/v1/learning_resources_search/?sortby=-views
http://localhost:8063/api/v1/learning_resources/?sortby=upcoming
http://localhost:8063/api/v1/learning_resources_search/?sortby=upcoming

- Go to http://localhost:8063/api/v1/schema/redoc verify that the documentation for sorby for api/v1/learning_resources and api/v1/learning_resources_search looks ok

- Go to http://localhost:8063/ and verify that the upcoming carusels are populated
